File: university-service/src/test/java/ua/com/foxminded/courseproject/repository/StudentRepositoryTest.java
package ua.com.foxminded.courseproject.repository;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.data.util.Streamable;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.transaction.annotation.Transactional;
import ua.com.foxminded.courseproject.entity.Group;
import ua.com.foxminded.courseproject.entity.Student;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.assertEquals;

@SpringBootTest
@Sql(value = "classpath:initial_data.sql")
@Transactional
class StudentRepositoryTest {

    @Autowired
    private StudentRepository repository;

    @Test
    void findAll_shouldReturnListStudents() {
        Integer expectedSize = 2;

        List<Student> Students = Streamable.of(repository.findAll()).toList();

        assertEquals(expectedSize, Students.size());
    }

    @Test
    void findById_shouldReturnStudent_whenIdExists() {
        UUID id = UUID.fromString("f92afb9e-462a-11ed-b878-0242ac120002");
        String expectedFirstname = "Yura";

        Student actual = repository.findById(id).get();

        assertEquals(expectedFirstname, actual.getFirstName());
    }

    @Test
    void findById_shouldReturnEmptyOptional_whenIdNotExists() {
        UUID id = UUID.fromString("e966f7c0-4621-11ed-b838-0242ac120002");

        Optional<Student> actual = repository.findById(id);

        assertEquals(true, actual.isEmpty());
    }

    @Test
    void save_shouldAddStudent_whenStudentNotExists() {
        String testStr = "test";
        Student expected = new Student();
        Group group = new Group();
        group.setId(UUID.fromString("4937378e-4620-11ed-b878-0242ac120002"));
        group.setName("ET-01");
        expected.setFirstName(testStr);
        expected.setLastName(testStr);
        expected.setGroup(group);
        expected.setBirthDay(LocalDate.now());
        expected.setCourse(4);
        expected.setCaptain(false);

        repository.save(expected);

        assertEquals(expected, repository.findById(expected.getId()).get());
    }

    @Test
    void save_shouldChangeStudent_whenStudentExists() {
        Student expected = Streamable.of(repository.findAll()).stream().findAny().get();
        expected.setLastName("test");

        repository.save(expected);

        assertEquals(expected, repository.findById(expected.getId()).get());
    }

    @Test
    void delete_shouldDeleteStudent_whenStudentExists() {
        Student Student = Streamable.of(repository.findAll()).stream().findAny().get();
        Integer expectedSize = 1;

        repository.delete(Student);
        List<Student> Students = Streamable.of(repository.findAll()).toList();

        assertEquals(expectedSize, Students.size());
    }

    @Test
    void delete_shouldNotDeleteStudents_whenStudentNotExists() {
        String testStr = "test";
        Student student = new Student();
        Group group = new Group();
        group.setId(UUID.fromString("4937378e-4620-11ed-b878-0242ac120002"));
        group.setName("ET-01");
        student.setFirstName(testStr);
        student.setLastName(testStr);
        student.setGroup(group);
        student.setBirthDay(LocalDate.now());
        student.setCourse(4);
        student.setCaptain(false);
        Integer expectedSize = 2;

        repository.delete(student);
        List<Student> Students = Streamable.of(repository.findAll()).toList();

        assertEquals(expectedSize, Students.size());
    }

}

File: university-service/src/test/kotlin/ua/com/foxminded/courseproject/repository/StudentRepositoryTest.kt
package ua.com.foxminded.courseproject.repository

import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest
import org.springframework.context.annotation.Import
import reactor.test.StepVerifier
import ua.com.foxminded.courseproject.config.DBTestConfig
import ua.com.foxminded.courseproject.entity.Group
import ua.com.foxminded.courseproject.entity.Student
import ua.com.foxminded.courseproject.mapper.GroupMapper
import ua.com.foxminded.courseproject.mapper.StudentMapper
import java.time.LocalDate
import java.util.*


@DataMongoTest
@Import(StudentRepositoryImp::class, StudentMapper::class, GroupMapper::class)
internal open class StudentRepositoryTest : DBTestConfig() {

    @Autowired
    private lateinit var repository: StudentRepository

    @Test
    fun findAll_shouldReturnFluxStudents() {
        val expectedSize = 2

        val students = repository.findAll()

        StepVerifier.create(students)
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()
    }

    @Test
    fun findById_shouldReturnStudent_whenIdExists() {
        val id = UUID.fromString("f92afb9e-462a-11ed-b878-0242ac120002")
        val expectedFirstname = "Yura"

        val actual = repository.findById(id)

        StepVerifier.create(actual)
            .expectNextMatches { expectedFirstname == it.firstName }
            .verifyComplete()
    }

    @Test
    fun findById_shouldReturnEmptyMono_whenIdNotExists() {
        val id = UUID.fromString("e966f7c0-4621-11ed-b838-0242ac120002")

        val actual = repository.findById(id)

        StepVerifier.create(actual)
            .expectNextCount(0)
            .verifyComplete()
    }

    @Test
    fun save_shouldAddStudent_whenStudentNotExists() {
        val testStr = "test"
        val expected = Student()
        val group = Group()
        group.id = UUID.fromString("4937378e-4620-11ed-b878-0242ac120002")
        group.name = "ET-01"
        expected.firstName = testStr
        expected.lastName = testStr
        expected.group = group
        expected.birthDay = LocalDate.now()
        expected.course = 4
        expected.captain = false

        repository.save(expected).block()


        StepVerifier.create(repository.findById(expected.id!!))
            .expectNextMatches { it.firstName == expected.firstName }
            .verifyComplete()

    }


    @Test
    fun save_shouldChangeStudent_whenStudentExists() {
        val expected = repository.findAll().blockFirst()
        expected.lastName = "test"

        repository.save(expected).block()

        expected.id?.let { repository.findById(it) }?.let {
            StepVerifier.create(it)
                .expectNextMatches { it.equals(expected) }
                .verifyComplete()
        }
    }

    @Test
    fun delete_shouldDeleteStudent_whenStudentExists() {
        val student = repository.findAll().blockFirst()
        val expectedSize = 1

        repository.delete(student).block()

        StepVerifier.create(repository.findAll())
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()
    }

    @Test
    fun delete_shouldNotDeleteStudents_whenStudentNotExists() {
        val testStr = "test"
        val student = Student()
        val group = Group()
        group.id = UUID.fromString("4937378e-4620-11ed-b878-0242ac120002")
        group.name = "ET-01"
        student.firstName = testStr
        student.lastName = testStr
        student.group = group
        student.birthDay = LocalDate.now()
        student.course = 4
        student.captain = false
        val expectedSize = 2

        repository.delete(student).block()

        StepVerifier.create(repository.findAll())
            .expectNextCount(expectedSize.toLong())
            .verifyComplete()
    }
}
